### PR TITLE
display multiple fields on one form

### PR DESCRIPTION
**Pull request by Ozrei**

### CODE REVIEW

File: keyform/forms.py [hunk 1]
```diff
@@ -13,3 +15,18 @@ def __init__(self, *args, **kwargs):
         #removes blank choices from Radio Select options
         self.fields['payment_method'] = TypedChoiceField(widget=RadioSelect(), choices=Request.PAYMENT_TYPES)
         self.fields['reason_for_request'] = TypedChoiceField(widget=RadioSelect(), choices=Request.REQUEST_TYPES)
+
+class OtherForm(forms.ModelForm):
```
Comment (enki211) [inline keyform/forms.py]: Probably should have a better name than OtherForm

File: keyform/templates/keyform/add_form.html [hunk 2]
```diff
@@ -6,15 +6,12 @@
...
-                            <p class="help">{{ field.help_text|safe }}</p>
-                        {% endif %}
-                    </div>
-                {% endfor %}
+                <div class="fieldWrapper">
+                    {{ form.as_table }}
+                    {{ keydata_formset.as_table }}
+                    {{ keydata_formset.management_form }}
+                </div>
+                <button id="add another">Add New Field</button>
```
Comment (enki211) [inline keyform/templates/keyform/add_form.html]: trans tag

Comment (enki211) [inline keyform/templates/keyform/add_form.html]:
```
add-another

File: keyform/views.py [hunk 3]
```diff
@@ -3,16 +3,41 @@
...
 class NewForm(CreateView):
     template_name = "keyform/add_form.html"
-    model = Request
+    model = Request, KeyData
     success_url = reverse_lazy('home')
-    form_class = CreateForm
+    form_class = RequestFormSet
+    def get_context_data(self, *args, **kwargs):
+        context = super(NewForm, self).get_context_data(*args, **kwargs)
+        from pprint import pprint
```
Comment (enki211) [inline keyform/views.py]: Why are you printing?

File: keyform/templates/keyform/add_form.html [hunk 4]
```diff
@@ -6,15 +6,12 @@
...
-                {% for field in form %}
-                    <div class="fieldWrapper">
-                        {{ field.errors }}
-                        {{ field.label_tag }} {{ field }}
-                        {% if field.help_text %}
-                            <p class="help">{{ field.help_text|safe }}</p>
-                        {% endif %}
-                    </div>
-                {% endfor %}
+                <div class="fieldWrapper">
```
Comment (enki211) [inline keyform/templates/keyform/add_form.html]: For html classes we usually do 
```
field-wrapper

File: keyform/static/keyform/css/base.css [hunk 5]
```diff
@@ -2,3 +2,16 @@ ul{
     list-style-type: none;
     margin: 0;
 }
+
+div.title-bar{
```
Comment (enki211) [inline keyform/static/keyform/css/base.css]: It does not look like you use this in your pull request

File: keyform/forms.py [hunk 6]
```diff
@@ -13,3 +15,18 @@ def __init__(self, *args, **kwargs):
...
+
+class OtherForm(forms.ModelForm):
+    class Meta:
+        model = KeyData
+        fields = ['new_core_number', 'key_type', 'room_number', 'lost_key_number', 'quantity']
+
+RequestFormSet = inlineformset_factory(Request, KeyData, extra=1, can_delete=False, exclude=[])
+
+
+class UserMultiForm(MultiModelForm):
```
Comment (enki211) [inline keyform/forms.py]: Probably should have a more descriptive name than `UserMultiForm`

**Comment by Ozrei** [comment]
To make the multiform display work:
pip install django-betterforms

File: keyform/templates/keyform/add_form.html [hunk 7]
```diff
@@ -6,15 +6,12 @@
...
-                            <p class="help">{{ field.help_text|safe }}</p>
-                        {% endif %}
-                    </div>
-                {% endfor %}
+                <div class="field-wrapper">
+                    {{ form.as_table }}
+                    {{ keydata_formset.as_table }}
+                    {{ keydata_formset.management_form }}
+                </div>
+                <button id="add-another">{%trans "Add New Field" %}</button>
```
Comment (enki211) [inline keyform/templates/keyform/add_form.html]: We may want to look into something more descriptive than `Add New Field`

File: keyform/views.py [hunk 8]
```diff
@@ -3,16 +3,37 @@
...
+def KeyRequest(request):
+    if request.POST:
+        form = CreateForm(request.POST)
+        if form.is_valid():
+            create_new = form.save(commit=False)
+            key_formset = RequestFormSet(request.POST, instance=create_new)
+            if key_formset.is_valid():
+                create_new.save()
+                key_formset.save()
+                return HttpResponseRedirect(reverse('create'))
```
Comment (enki211) [inline keyform/views.py]: If it is valid and they correctly submitted it should we send them to the home page instead?

File: keyform/views.py [hunk 9]
```diff
@@ -3,16 +3,37 @@
...
 from django.urls import reverse_lazy
+from django.shortcuts import render
+from django.http import HttpResponseRedirect
+from django.core.urlresolvers import reverse
 
 class HomeView(ListView):
     model = Request
     template_name = "keyform/home.html"
 
 class NewForm(CreateView):
```
Comment (enki211) [inline keyform/views.py]: Is the NewForm class used anywhere?

File: keyform/urls.py [hunk 10]
```diff
@@ -4,5 +4,5 @@
 
 urlpatterns = [
     url(r'^$', views.HomeView.as_view(), name='home'),
-    url(r'^create_new', views.NewForm.as_view(), name='create')
+    url(r'^create_new', views.KeyRequest, name='create')
```
Comment (enki211) [inline keyform/urls.py]: You may want to throw a $ after create_new
also maybe a different url?
```python
url(r'^create_new$', views.KeyRequest, name='create'),

**Comment by enki211** [comment]
You may not need the betterforms thing

File: keyform/templates/keyform/add_form.html [hunk 11]
```diff
@@ -6,15 +6,12 @@
...
-                            <p class="help">{{ field.help_text|safe }}</p>
-                        {% endif %}
-                    </div>
-                {% endfor %}
+                <div class="field-wrapper">
+                    {{ form.as_table }}
+                    {{ keydata_formset.as_table }}
+                    {{ keydata_formset.management_form }}
+                </div>
+                <button id="add-another">{%trans "Add Key Type Field" %}</button>
```
Comment (enki211) [inline keyform/templates/keyform/add_form.html]: What does Add Key Type field mean?

File: keyform/static/keyform/js/libraries/add_form.js [hunk 12]
```diff
@@ -0,0 +1,7 @@
+$(document).ready(function() {
```
Comment (enki211) [inline keyform/static/keyform/js/libraries/add_form.js]: Why is this file a duplicate and in the libraries folder?

File: keyform/static/keyform/js/add_form.js [hunk 13]
```diff
@@ -0,0 +1,7 @@
+$(document).ready(function() {
+    $('#generate_forms').click(function() {
+        var form_idx = $('#id_keydata_set-TOTAL_FORMS').val();
+        $('#form_set').append($('#empty_form').html().replace(/__prefix__/g, form_idx));
+        $('#id_keydata_set-TOTAL_FORMS').val(parseInt(form_idx) + 1);
+    });
+});
```
Comment (enki211) [inline keyform/static/keyform/js/add_form.js]: Newline

File: keyform/static/keyform/css/add_form.css [hunk 14]
```diff
@@ -0,0 +1,5 @@
+div.keydata_form {
+    border-style: solid;
+    border-width: 1px;
+    border-radius: 5px;
+}
```
Comment (enki211) [inline keyform/static/keyform/css/add_form.css]: Newline

File: keyform/static/keyform/css/base.css [hunk 15]
```diff
@@ -0,0 +1,17 @@
+ul{
+    list-style-type: none;
+    margin: 0;
+}
+
+div.title-bar{
+    background-color: maroon;
```
Comment (enki211) [inline keyform/static/keyform/css/base.css]: If you are going for the MSU color I believe it is `#5E0009`

File: keyform/static/keyform/css/add_form.css [hunk 16]
```diff
@@ -0,0 +1,5 @@
+div.keydata_form {
```
Comment (enki211) [inline keyform/static/keyform/css/add_form.css]: `keydata-form`

File: keyform/static/keyform/js/add_form.js [hunk 17]
```diff
@@ -0,0 +1,7 @@
+$(document).ready(function() {
```
Comment (enki211) [inline keyform/static/keyform/js/add_form.js]: It does not look like this file is included on the page

File: keyform/static/keyform/js/add_form.js [hunk 18]
```diff
@@ -0,0 +1,7 @@
...
+    $('#generate_forms').click(function() {
```
Comment (enki211) [inline keyform/static/keyform/js/add_form.js]: Where is the tag with id `generate_forms`?

File: keyform/templates/keyform/add_form.html [hunk 19]
```diff
@@ -1,20 +1,35 @@
...
-                    <div class="fieldWrapper">
-                        {{ field.errors }}
-                        {{ field.label_tag }} {{ field }}
-                        {% if field.help_text %}
-                            <p class="help">{{ field.help_text|safe }}</p>
-                        {% endif %}
+                <div class="field-wrapper">
+                    {{ form.as_p }}
+                    {{ keydata_formset.management_form }}
+                    <div id="form_set">
```
Comment (enki211) [inline keyform/templates/keyform/add_form.html]: `form-set`

File: keyform/templates/keyform/add_form.html [hunk 20]
```diff
@@ -1,20 +1,35 @@
...
+                        {% for form in keydata_formset.forms %}
+                            <div class="keydata_form">
```
Comment (enki211) [inline keyform/templates/keyform/add_form.html]: `keydata-form`

File: keyform/templates/keyform/add_form.html [hunk 21]
```diff
@@ -1,20 +1,39 @@
 {% extends "keyform/base.html" %}
+{% load static %}
 {% load i18n %}
 
+{% block javascript %}
+{{ block.super }}
+<script type="text/javascript" src="{% static 'keyform/js/add_form.js' %}"></script>
```
Comment (enki211) [inline keyform/templates/keyform/add_form.html]: indentation

File: keyform/templates/keyform/add_form.html [hunk 22]
```diff
@@ -1,20 +1,39 @@
...
+                    {{ form.as_p }}
+                    {{ keydata_formset.management_form }}
+                    <div id="form-set">
+                        {% for form in keydata_formset.forms %}
+                            <div class="keydata-form">
+                                {{ form.as_p }}
+                            </div>
+                        {% endfor %}
+                    </div>
+                    <input type="button" value="Add Another Key" id="generate_forms">
```
Comment (enki211) [inline keyform/templates/keyform/add_form.html]: trans tag around `Add Another Key`, close out the input `/>`

File: keyform/views.py [hunk 23]
```diff
@@ -3,16 +3,40 @@
...
+                return HttpResponseRedirect(reverse('home'))
+    else:
+        form = CreateForm()
+        key_formset = RequestFormSet(instance=Request())
+    return render(request, "keyform/add_form.html", {
+        "form": form,
+        "keydata_formset": key_formset,
+    })
+
+def myview(request):
```
Comment (enki211) [inline keyform/views.py]: I don't think this is needed

File: keyform/templates/keyform/add_form.html [hunk 24]
```diff
@@ -1,20 +1,39 @@
...
+                    <div id="empty_form" style="display:none">
+                        <div class="keydata-form">
+                            {{ keydata_formset.empty_form.as_p }}
+                        </div>
                     </div>
-                {% endfor %}
+                    <div id="forms"></div>
```
Comment (enki211) [inline keyform/templates/keyform/add_form.html]: Is this div necessary?

File: keyform/static/keyform/css/add_form.css [hunk 25]
```diff
@@ -0,0 +1,5 @@
+div.keydata-form {
+    border-style: solid;
+    border-width: 1px;
+    border-radius: 5px;
```
Comment (shaneallgeier) [inline keyform/static/keyform/css/add_form.css]: This is kinda ugly but I guess it will work until the redesign happens

File: keyform/views.py [hunk 26]
```diff
@@ -3,16 +3,31 @@
...
 class HomeView(ListView):
     model = Request
     template_name = "keyform/home.html"
 
-class NewForm(CreateView):
-    template_name = "keyform/add_form.html"
-    model = Request
-    success_url = reverse_lazy('home')
-    form_class = CreateForm
+def KeyRequest(request):
```
Comment (shaneallgeier) [inline keyform/views.py]: According to [PEP8](https://www.python.org/dev/peps/pep-0008/#function-names), function names should be lowercase, with words separated by underscores.

File: keyform/static/keyform/css/base.css [hunk 27]
```diff
@@ -1,4 +1,17 @@
...
+ul{
+    list-style-type: none;
+    margin: 0;
+}
+
+div.title-bar{
+    background-color: #5E0009;
+}
+
+div.field-wrapper{
```
Comment (shaneallgeier) [inline keyform/static/keyform/css/base.css]: This should be moved to add_form.css

File: keyform/static/keyform/css/base.css [hunk 28]
```diff
@@ -1,4 +1,17 @@
-ul{
-    list-style-type: none;
-    margin: 0;
-}
+ul{
+    list-style-type: none;
+    margin: 0;
+}
+
+div.title-bar{
```
Comment (shaneallgeier) [inline keyform/static/keyform/css/base.css]: This should be removed since this change is unrelated to this pull request.

File: keyform/forms.py [hunk 29]
```diff
@@ -13,3 +14,10 @@ def __init__(self, *args, **kwargs):
         #removes blank choices from Radio Select options
         self.fields['payment_method'] = TypedChoiceField(widget=RadioSelect(), choices=Request.PAYMENT_TYPES)
         self.fields['reason_for_request'] = TypedChoiceField(widget=RadioSelect(), choices=Request.REQUEST_TYPES)
+
+class KeyDataForm(forms.ModelForm):
```
Comment (shaneallgeier) [inline keyform/forms.py]: This class is completely unused.

File: keyform/views.py [hunk 30]
```diff
@@ -3,16 +3,31 @@
 
 from django.views.generic import CreateView
 from django.views.generic.list import ListView
-from keyform.forms import CreateForm
-from keyform.models import Request
+from keyform.forms import CreateForm, KeyDataForm, RequestFormSet
+from keyform.models import Request, KeyData
```
Comment (shaneallgeier) [inline keyform/views.py]: KeyData is unused in this file

**Comment by enki211** [comment]
Please update with master

File: keyform/views.py [hunk 31]
```diff
@@ -3,16 +3,31 @@
...
 class HomeView(ListView):
     model = Request
     template_name = "keyform/home.html"
 
-class NewForm(CreateView):
-    template_name = "keyform/add_form.html"
-    model = Request
-    success_url = reverse_lazy('home')
-    form_class = CreateForm
+def key_request(request):
```
Comment (shaneallgeier) [inline keyform/views.py]: You're going to hate me for this: Make this into a class-based view that inherits from TemplateView

Comment (Ozrei) [inline keyform/views.py]: 你是餐巾

**Comment by enki211** [comment]
A delete button that would remove keydata forms would be nice

File: keyform/static/keyform/js/add_form.js [hunk 32]
```diff
@@ -4,4 +4,9 @@ $(document).ready(function() {
         $('#form-set').append($('#empty_form').html().replace(/__prefix__/g, form_idx));
         $('#id_keydata_set-TOTAL_FORMS').val(parseInt(form_idx) + 1);
     });
+
+    $('#delete_button').click(function() {
+        if ($('#form-set .keydata-form').length > 1)
+            $('#form-set .keydata-form:last-child').remove();
```
Comment (enki211) [inline keyform/static/keyform/js/add_form.js]: Please put this in braces

File: keyform/templates/keyform/add_form.html [hunk 33]
```diff
@@ -18,18 +18,19 @@
...
-                        {% for form in keydata_formset.forms %}
+                        {% for keydata_form in form.request_formset.forms %}
                             <div class="keydata-form">
-                                {{ form.as_p }}
+                                {{ keydata_form.as_p }}
                             </div>
                         {% endfor %}
                     </div>
                     <input type="button" value="{% trans 'Add Another Key' %}" id="generate_forms"/>
+                    <input type="button" value="{% trans 'Cancel' %}" id="delete_button"/>
```
Comment (enki211) [inline keyform/templates/keyform/add_form.html]: `delete-button`

File: keyform/views.py [hunk 34]
```diff
@@ -1,33 +1,36 @@
 # -*- coding: utf-8 -*-
 from __future__ import unicode_literals
 
-from django.views.generic import CreateView
+from django.views.generic import FormView
 from django.views.generic.list import ListView
 from keyform.forms import CreateForm, RequestFormSet
 from keyform.models import Request
 from django.urls import reverse_lazy
```
Comment (enki211) [inline keyform/views.py]: Is this `reverse_lazy` no longer needed?

File: keyform/views.py [hunk 35]
```diff
@@ -1,33 +1,36 @@
...
 from django.shortcuts import render
```
Comment (enki211) [inline keyform/views.py]: Is `render` needed?

**Comment by enki211** [comment]
Can you add a link to the create page on the titlebar?

File: keyform/static/keyform/js/add_form.js [hunk 36]
```diff
@@ -0,0 +1,14 @@
...
+    $('#generate_forms').click(function() {
+        var form_idx = $('#id_keydata_set-TOTAL_FORMS').val();
+        $('#form-set').append($('#empty_form').html().replace(/__prefix__/g, form_idx));
+        
+    });
+
+    $('#delete-button').click(function() {
+        if ($('#form-set .keydata-form').length > 1){
+            $('#form-set .keydata-form:last-child').remove();
+            $('#id_keydata_set-TOTAL_FORMS').val(parseInt(form_idx) - 1);
```
Comment (shaneallgeier) [inline keyform/static/keyform/js/add_form.js]: This line is unnecessary.
